ScopedGraphCapture

mbezuljTT opened the pull request:
### Problem description
If one wants to catch exception during graph capture thrown by op, they need to end graph capture if it was still active. Requires manual tracking.

### What's changed
Introduced ScopedGraphCapture, a wrapper class that begins capture on the creation, can be manually ended, or ends when goes out of scope. 

Utilized in the op_constraints 

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/12757393308/job/35568824855

mbezuljTT committed:
GraphCaptureScopeGuard

ayerofieiev-tt commented on ttnn/cpp/ttnn/graph/graph_query_op_constraints.hpp:
```diff
@@ -44,30 +44,30 @@ struct QueryResponse {
-        GraphProcessor::begin_graph_capture(GraphProcessor::RunMode::NO_DISPATCH);
-        num_of_active_graph_captures++;
+        {
+            auto capture_outer = GraphCaptureScopeGuard(GraphProcessor::RunMode::NO_DISPATCH);
```
Interesting approach, I thought `GraphProcessor::begin_graph_capture` will return a guard instead. 
One concern I have is that `GraphCaptureScopeGuard` does not make it clear that you just started tracing.
It is not a major one. Wonder what @dmakoviichuk-tt thinks.

dmakoviichuk-tt replied:
It sounds ok for me. But btw if exception was throw it means we didn;t close. some functions. is it ok?

bbradelTT replied:
Maybe changing the name to GraphCapture, ScopedGraphCapture, or ScopeGuardedGraphCapture might help. Thoughts?

ayerofieiev-tt reviewed (commented):
Please add a test

dmakoviichuk-tt reviewed (approved):
Looks good but please add some tests.
You can make a few fake ops with nested calls and throw something from inside.

dmakoviichuk-tt commented:
maybe it is a good idea to call all end_function for all funcrtions which we didn't close.

mbezuljTT committed:
added GraphCaptureScopeGuard test

mbezuljTT commented:
I've rebased and added a test. I didn't make a fake op, I've caused existing op (softmax) to fail by providing wrong arguments.  

> maybe it is a good idea to call all end_function for all funcrtions which we didn't close, it might happen in your case. But lest take a look at test.

as you said, capture that had exception thrown - it's functions would not be ended. it doesn't impact subsequent graph captures. 

this is the graph with exception 
<img width="633" alt="image" src="https://github.com/user-attachments/assets/6f3ed133-4ce4-4760-90a4-3dd0d16f0922" />

while these are before/after
<img width="1577" alt="image" src="https://github.com/user-attachments/assets/bbddba6a-46da-4dad-beef-d7bbc48d3fe6" />

this is ok for compiler scenario, but I don't have all context on the graph capture usage; @dmakoviichuk-tt  please let me know what is your preference - to insert function ends or keep it like this.

mbezuljTT committed:
GraphCaptureScopedGuard->ScopedGraphCapture

mbezuljTT committed:
fix

mbezuljTT committed:
remove exception in the operation #1 because TT_ASSERT is ignored in Release build
